nipype2/engine/submiter.py

```diff
@@ -19,62 +19,61 @@
         if self.plugin == "mp":
-            self.worker = MpWorker(done=self.done)
+            self.worker = MpWorker()
         logger.debug('Initialize Submitter, graph: {}'.format(graph))
-        self._count_subm = 0
-        self._count_done = 0
-
+        self._to_finish = list(self.graph)
 
     def run_workflow(self):
         for (i_n, node) in enumerate(self.graph):
-            # checking if a node has all input (doesnt have to wait for others)
+            # submitting all the nodes who are self sufficient (self.graph is already sorted)
             if node.sufficient:
-                self._count_subm += 1
                 self.submit_work(node)
             # if its not, its been added to a line
             else:
-                self.node_line.append(node)
-
-        # TODO: should add an extra condition to not stay here forever
-        while self.node_line:
-            logger.debug("Submitter, node_line BEFORE trying to get new output: {}".format(self.node_line))
-            try:
-                el_done = self.done.get(timeout=1)
-                logger.debug("Submitter, el from self.done: {}".format(el_done))
-                self._count_done += 1
-                time.sleep(2)
-                self.connecting_output(el_done)
-            except queue.Empty:
-                time.sleep(3)
-            logger.debug("Submitter, node_line AFTER trying to get new output: {}".format(self.node_line))
-
-        self._collecting_results()
-
-
-    def connecting_output(self, el_out):
-        from_node = self.graph_names[el_out[2]]#el_out[2]
-        for (from_socket, to_node, to_socket) in from_node.sending_output:
-            if (from_node, from_socket, to_socket) in to_node.needed_outputs:
-                # TODO this works only because there is no mapper!
-                file_output = [name for name in glob.glob("{}/*/{}.txt".format(from_node.nodedir, from_socket))][0]
-                with open(file_output) as f:
-                    to_node.inputs.update({to_socket: eval(f.readline())})
-                to_node.needed_outputs.remove((from_node, from_socket, to_socket))
-
-                if not to_node.needed_outputs:
-                    self.node_line.remove(to_node)
-                    to_node.sufficient = True
-                    self._count_subm += 1
-                    self.submit_work(to_node)
-            else:
-                raise Exception("something wrong with connections")
+                break
+
+        # all nodes that are not self sufficient will go to the line
+        # (i think ordered list work well here, since it's more efficient to check within a specific order)
+        self.node_line = self.graph[i_n:]
+
+        # this parts submits nodes that are waiting to be run
+        # it should stop when nothing is waiting
+        while self._nodes_check():
+            logger.debug("Submitter, in while, node_line: {}".format(self.node_line))
+            time.sleep(3)
```

Comment:
why sleep here?

Reply:
ah i see, but i think the next while can potentially be combined with this while.

Reply:
was thinking about it, but decided that it will be easier to read at least for now. The first while is for   nodes which are waiting to start (so nothing goes here if there is no connect) and the second while simply waits for all nodes to finish. But will think about combining them.